_posts/2016-04-04-code-smells-part-II.md

```diff
@@ -0,0 +1,91 @@
+image:
+    src: /assets/img/custom/blog/code-smells_II.jpg
+    attribution:
+        text: Via Flickr/Creative Commons/Brian Fitzgerald (Creative Commons - Attribution-ShareAlike 2.0 Generic license)
+        href: https://www.flickr.com/photos/brian-fitzgerald/3334353375
+tags:
+- object-orientation abusers
+- change preventers
+- code smells
+- refactoring
+
+---
+In the last post, [Code Smells - Part I](http://codurance.com/2016/03/17/code-smells-part-I),  I talked about the bloaters: they are code smells that can be identified as Long Methods, Large Classes, Primitive Obsessions, Long Parameter List and Data Clumps. In this one, I would like to dig into the **Object-Orientation Abusers** and the **Change Preventers**. 
+
+## Object-Orientation Abusers
+
+This type of code smell usually happens when object-oriented principles are incomplete or incorrectly applied.
+
+
+#### Switch Statements
+
+This case is simple to identify: we have a switch case. But you should consider it a smell too if you find a sequence of ifs. (that's a switch case in disguise).
+Why are switch statements bad? Because when a new condition is added, you have to find every occurrence of that switch case. 
+So while talking to [David](https://twitter.com/DHatanian), he asked me: and what happens if I encapsulate the switch into a method, is it acceptable then? That's really a good question... If your switch case is only used to "take care" of one behaviour and that's it, then it might be ok . Remember identifying a code smell doesn't mean that you have to get always ride of it: it's a trade of. If you find your switch statement replicated and each replication has  different behaviour, then you cannot simply isolate the switch statement in a method. You need to find a proper "home" for it to be in. As a rule of thumb, you should think of polymorphism when you find yourself in this situation. There are two refactoring techniques that we can apply here: 
+
+ - **_Replace Type Code with Subclasses_**
+   This technique consists of creating subclasses for each switch case and applying the respective behaviour to these subclasses.
+ - **_Replace Type Code With Strategy_**
+   Similar to the above one, in this case, you should make use of one of the patterns: [State](https://en.wikipedia.org/wiki/State_pattern) or [Strategy](https://en.wikipedia.org/wiki/Strategy_pattern).
+
+So when to use one or the other? If the **_Type Code_** does not change the behaviour of a class you can use the **_Subclasses_** technique. Separating each behaviour in it's appropriated subclass will enforce the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) and make the code more readable in general. If you need to add another case, you just add a new class to your code without having to modify any other code. So you apply the [Open/Close Principle](https://en.wikipedia.org/wiki/Open/closed_principle).
```

Comment:
if there is no changed behaviour, what do we separate into the appropriate subclass?